Documentation revisions for v3

alawibaba opened the pull request:
This PR consists of documentation changes that I am suggesting as part of our v3 release.

Some elements (such as links to repos where the v3 binaries will be distributed) are only guesses at this point. We also expect these to be living documents, so users are encouraged to refer to the hosted version of the documents when possible (geopm.github.io).

I will try to indicate where I need another set of eyes.

alawibaba commented on service/docs/source/admin.rst:
Could someone please validate the file locations? Maybe @bgeltz ?

dannosliwcd commented on service/docs/source/admin.rst:
```diff
@@ -16,42 +15,45 @@ Service.  Additional information is available on other pages:
+may generate and modify additional files during its active state. These files
+are housed within two primary directories:
+
+- ``/etc/geopm``: This directory contains access control lists. Files here
+  persist across both reboots and service restarts.
```
More generally, it contains any config files. Other than access control config, MSRIOGroup configs and ConstConfigIOGroup configs come to mind. The service does come with MSRIOGroup configs preinstalled, but additional ones could be configured by an admin.

dannosliwcd commented on service/docs/source/overview.rst:
```diff
@@ -309,33 +312,31 @@ Reading Signals
-The telemetry that is output from ``geopmread`` or the APIs will be
-automatically aggregated based on the requested domain and the aggregation
+The telemetry that is output from ``geopmread`` or the APIs will automatically
+be aggregated based on the requested domain and the aggregation
 type.
```
We should probably add a second example with a different agg type so people understand what we mean by this. E.g., CPU_ENERGY uses a sum aggregator vs the time-weighted mean used by CPU_FREQEUNCY_STATUS.

The second example doesn't need as much detail as the first example's discussion. We just need to highlight that it is aggregated differently.

dannosliwcd commented on service/docs/source/overview.rst:
```diff
@@ -309,33 +312,31 @@ Reading Signals
-from code.
+To fetch platform telemetry and output it to the console or a file:
+
+- From the command-line: Use `geopmsession`. Its input arguments are similar to `geopmread`.
```
> Its input arguments are similar to `geopmread`

I don't think that is true. The way `geopmsession` interprets `stdin` is similar to the way `geopmread` interprets its positional args.

alawibaba commented:
@dannosliwcd I addressed your comments; thank you!

avilcheslopez commented on service/docs/source/admin.rst:
```diff
@@ -2,12 +2,11 @@
-created and modified by the GEOPM Service, how the files in those
-directories are used, and a command line tool to configure the GEOPM
-Service.  Additional information is available on other pages:
+This guide offers insights into aspects of the GEOPM Service relevant for
```
Very minor nit: for brevity, you could remove this first line (and make very minor adjustments in the following lines), as this is implied by the section title.

avilcheslopez commented on service/docs/source/admin.rst:
```diff
@@ -16,42 +15,45 @@ Service.  Additional information is available on other pages:
+  control lists. Files here persist across both reboots and service restarts.
+
+- ``/run/geopm``: Files here monitor data about clients actively engaging the
+  service. They help maintain the GEOPM Service's state. Should the service
```
Minor nit: I would suggest a minor rewording to make it clear that you're talking about different kinds of files being maintained in this directory. The way it's worded in the first two sentences, it seems like you're talking about one kind of file (which maintains client data and service state).

Also, I believe the files needed by the save/restore mechanism are stored in this directory.

avilcheslopez commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,18 @@
 
-Guide to Source Builds
-======================
+This documentation provides build instructions for administrators who are interested in installing the GEOPM Service based on a checkout from the GEOPM git repository. Following these instructions will assist users in creating RPM packages compatible with various Linux distributions.
```
Maintain the column limit (more instances below). We haven't been very strict about the limit in these documentation files. Some use a limit around 70, others around 80. Most files try to stay around 70.

avilcheslopez commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,18 @@
-RPM packaging system enables some cross-Linux-distribution
-compatibility for the installation of the systemd service, and
-provides a clean mechanism for uninstall.
+We have tested packages created through this process on CentOS 8, RHEL 8, and SLES 15 SP2 Linux distributions. Feedback from tests on other distributions such as Fedora, openSUSE Leap, and Ubuntu is highly appreciated. 
```
Minor suggestion:
```suggestion
We have tested packages created through this process on CentOS 8, RHEL 8, and SLES 15 SP2 Linux distributions. Feedback from installations on other distributions such as Fedora, openSUSE Leap, and Ubuntu would be greatly appreciated. 
```

avilcheslopez commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,18 @@
-provides a clean mechanism for uninstall.
+We have tested packages created through this process on CentOS 8, RHEL 8, and SLES 15 SP2 Linux distributions. Feedback from tests on other distributions such as Fedora, openSUSE Leap, and Ubuntu is highly appreciated. 
 
+Please note that currently, there is no .deb support. GEOPM Service installation is reliant on the RPM packaging system. RPM provides cross-Linux-distribution compatibility for the installation of the systemd service and facilitates a clean uninstallation process.
```
```suggestion
Please note that currently, there is no Debian support. GEOPM Service installation is reliant on the RPM packaging system. RPM provides cross-Linux-distribution compatibility for the installation of the systemd service and facilitates a clean uninstallation process.
```

or `...Debian package...`?

cmcantalupo replied:
Um...we have debian support!  There is even a ``make deb`` target that will build the package files.

avilcheslopez commented on service/docs/source/build.rst:
```diff
@@ -45,37 +32,26 @@ directory:
-After following the instructions above, install the RPM files by
-executing the ``install_service.sh`` script located in the
-``service/integration`` sub-directory of the GEOPM repository.
+Post following the aforementioned instructions, install the RPM files by executing the `install_service.sh` script located within `service/integration` subdirectory of the GEOPM repository.
```
Column limit. More instances in this file.

avilcheslopez commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,18 @@
 
-Support for packaging for CentOS 8, RHEL 8, and SLES 15 SP2 is provided
-by the GEOPM service build system.
+The GEOPM service build system provides support for packaging for CentOS 8, RHEL 8, and SLES 15 SP2.
```
Minor suggestion:

```suggestion
The GEOPM service build system provides packaging support for CentOS 8, RHEL 8, and SLES 15 SP2.
```

avilcheslopez commented on service/docs/source/client.rst:
```diff
@@ -1,21 +1,19 @@
+utilize the PlatformIO capabilities through different interfaces
+such as C, C++, Python, and command line tools. The system 
+administrator can manage user access rights using the
+``geopmaccess`` command line tool. The PlatformTopo
```
```suggestion
``geopmaccess`` command line tool. PlatformTopo
```

avilcheslopez commented on service/docs/source/devel.rst:
```diff
@@ -1,197 +1,98 @@
-Prior to following the build steps, it's recommended to get familiar with
-the build requirements for the :doc:`GEOPM service<requires>` and the
-:doc:`GEOPM HPC runtime<runtime>`.
+If you wish to modify the source code in the GEOPM git repository, this guide provides instructions for the process. The GEOPM repository utilizes two separate autotools-based build systems, used for compiling, testing, and installing software components designed with C++ & Python. 
```
Column limit (more instances below)

avilcheslopez replied:
```suggestion
If you wish to modify the source code in the GEOPM git repository, this guide provides instructions for the process. The GEOPM repository utilizes two separate autotools-based build systems, used for compiling, testing, and installing software components designed with C++ and Python. 
```

avilcheslopez commented on service/docs/source/devel.rst:
```diff
@@ -1,197 +1,98 @@
-the configure option is specified as above. then the following
-modifications to the user's environment should be made prior to
-running any GEOPM tools:
+When building from source and configured with the `--prefix` option, the libraries, binaries, and python tools will not install into the standard system paths. At this point, you must modify your environment to specify the installed location. 
```
```suggestion
When building from source and configured with the `--prefix` option, the libraries, binaries, and Python tools will not install into the standard system paths. At this point, you must modify your environment to specify the installed location. 
```

More instances elsewhere.

avilcheslopez commented on service/docs/source/index.rst:
```diff
@@ -4,28 +4,24 @@ Welcome to GEOPM
+The Global Extensible Open Power Manager (GEOPM) serves as a framework for investigating 
+energy and power optimizations geared towards heterogeneous platforms. It offers a whole 
+range of built-in functions. A basic use case could be reading hardware counters and 
+configuring hardware controls using a command line tool on a specific compute node with platform independent syntax. 
```
Column limit

avilcheslopez commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,49 @@
-install any of the other packages mentioned in the guides linked
-above.  Installing the ``level-zero-devel`` package will enable the
-user to build GEOPM from source with GPU support if that is desired.
+Each time specified branches hosted on GitHub are updated, continuous integration generates packages for numerous Linux distributions. This process uses GitHub Actions, combined with the OpenSUSE Build Service (OBS). Here are the packages curated by OBS:
```
Column limit (more instances below)

avilcheslopez commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,49 @@
+Besides these routinely curated packages, the download repositories also grant access to ``python3-dasbus`` version 1.6. This version is a prerequisite for ``python3-geopmdpy``, but not included by the Linux distributions that back us.
+
+Installation of ``geopm-service`` will involve the dependency packages: ``libgeopmd2``, ``python3-geopmdpy`` and ``python3-dasbus``. However, explicit installation is obligatory for anyone seeking to use ``geopm-service-devel``.
+
```
```suggestion
```

avilcheslopez commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,49 @@
+   - Refer links provided
+
+Use the provided links to install GEOPM Service packages on your system. The links will also guide you through instructions for each supported operating system. If you add both the development repository and the release repository to your package configuration, you will always receive updates from the development branch. 
+
```
```suggestion
```

avilcheslopez commented on service/docs/source/requires.rst:
```diff
@@ -44,105 +39,82 @@ Upstream SLES and OpenSUSE Package Requirements
-The GEOPM service requires a more recent version of dasbus than is
-currently packaged by Linux distributions (dasbus version 1.5 or more
-recent).  The script located in the subdirectory of the GEOPM repo:
+A more recent version of dasbus (version 1.5 or later) is required for the GEOPM service.
```
```suggestion
A more recent version of Dasbus (version 1.5 or later) is required for the GEOPM service.
```

More instances throughout.

avilcheslopez commented on service/docs/source/requires.rst:
```diff
@@ -44,105 +39,82 @@ Upstream SLES and OpenSUSE Package Requirements
+It's distributed with OpenHPC and can be installed from the RPMs provided there.
+
+In the absence of both the msr-safe kernel driver and the GEOPM Systemd Service, root users may access MSRs
+using the standard msr driver. This can be loaded with the command:
```
```suggestion
using the standard MSR driver. This can be loaded with the command:
```

avilcheslopez commented on service/docs/source/requires.rst:
```diff
@@ -44,105 +39,82 @@ Upstream SLES and OpenSUSE Package Requirements
+improving the performance of GEOPM HPC runtime or other MSR usages. 
+
+Secondly, the msr-safe kernel driver enables user-level read and write operations of 
+the model-specific registers (MSRs) with access controlled by the system administrator.
```
Recommendation: move the term `model-specific registers` to earlier in the document when it's first used, use MSR from there on.

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -13,97 +13,56 @@ The architecture is designed to provide a secure infrastructure to
-longer run times which incur even higher energy costs per unit of work
-than if an adaptive algorithm were not applied.
-
+For hardware tuning algorithms, the first challenge involves generating a durable estimate of application performance. In energy efficiency terms, performance measurements are power-to-performance ratios, for example, "perf per watt". Therefore, any dynamic hardware power control tuning that aims for energy efficiency must formulate an estimate of application performance. Without specific applica ...
```
Column limit (more instances below)

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -13,97 +13,56 @@ The architecture is designed to provide a secure infrastructure to
-of the processes running the control algorithm reduces software
-security audit requirements significantly.
-
+The third challenge is to provide a software development platform suitable for control algorithms that can be securely deployed and use high-level languages for data analysis. To effectively enhance energy efficiency for certain applications, substantial software dependencies may be required. Control algorithms might lean on optimization software like machine learning packages or other numerical  ...
```
```suggestion
The third challenge is to provide a software development platform suitable for control algorithms that can be securely deployed and use high-level languages for data analysis. To effectively enhance energy efficiency for certain applications, substantial software dependencies may be required. Control algorithms might lean on optimization software like machine learning or other numerical packages. The application being optimized could include millions of lines of code, and there may be significant interlinking between the application and the control algorithm. Limiting the privileges of the process running the control algorithm significantly reduces software security audit requirements.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -13,97 +13,56 @@ The architecture is designed to provide a secure infrastructure to
 .. figure:: https://geopm.github.io/images/geopm-runtime-usage.svg
-   :alt: An illustration of geopmlaunch running on 2 servers, generating a
-         trace file per host, and one report across all hosts.
+   :alt: An illustration of geopmlaunch running on 2 servers, generating a trace file per host, and one report across all hosts.
```
```suggestion
   :alt: An illustration of ``geopmlaunch`` running on 2 servers, generating a trace file per host, and one report across all hosts.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -13,97 +13,56 @@ The architecture is designed to provide a secure infrastructure to
-   profiling and optimizing MPI applications. It
-   wraps a launcher application (srun in this example), generates a summarizing
-   report file, and optionally generates a time-series trace per host.
+   The ``geopmlaunch`` tool is the recommended user interface for GEOPM Runtime. It wraps a launcher application
+   (like srun in this example), generates a summarizing report file, and optionally generates a time-series trace for each host.
```
```suggestion
   The ``geopmlaunch`` tool is the recommended user interface for the GEOPM Runtime. It wraps a launcher application
   (like ``srun`` in this example), generates a summarizing report file, and optionally generates a time-series trace for each host.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-variables used by the GEOPM runtime that are otherwise controlled by the
-wrapper script.
+
+The GEOPM Runtime package installs the ``geopmlaunch`` command. This command is a wrapper for MPI launch commands such as ``srun``, ``aprun``, and ``mpiexec``, where the wrapper script enables the GEOPM runtime. The geopmlaunch command supports the same command-line interface as the underlying launch command, while extending the interface with GEOPM-specific options. The ``geopmlaunch`` applicati ...
```
```suggestion
The GEOPM Runtime package installs the ``geopmlaunch`` command. This command is a wrapper for MPI launch commands such as ``srun``, ``aprun``, and ``mpiexec``, where the wrapper script enables the GEOPM runtime. The ``geopmlaunch`` command supports the same command-line interface as the underlying launch command, while extending the interface with GEOPM-specific options. The ``geopmlaunch`` application launches the primary compute application and the GEOPM control thread on each compute node, and manages all process CPU affinity requirements. This wrapper is documented in the :doc:`geopmlaunch(1)<geopmlaunch.1>` man page.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```suggestion
GEOPM Runtime offers several built-in algorithms, each incorporated within an "Agent" implementing the :doc:`geopm::Agent(3) <GEOPM_CXX_MAN_Agent.3>` class interface. Developers can expand these algorithm features by creating an Agent plugin. An implementation of this class can be dynamically loaded at runtime by the GEOPM Controller. The Agent class determines what data is collected, how control decisions are made, and how messages are exchanged between Agents in the compute nodes' tree hierarchy. The GEOPM Service package, which resides in the service directory of the GEOPM repository, provides the PlatformIO interface which abstracts reading signals and writing controls from the Agent within a compute node. This allows Agent implementations to be ported to various hardware platforms without modification.
```

avilcheslopez commented on service/docs/source/runtime.rst:
Minor recommendation:

```suggestion
When building the GEOPM Runtime from source, additional requirements must be met. Those uninterested in building the GEOPM Runtime can ignore these requirements, or by providing the disable flag to the configure command line, users may skip particular GEOPM Runtime features enabled by these requirements.
```

In general, we're moving away from using the term "base" build/directory in favor of the term runtime.

avilcheslopez commented on service/docs/source/runtime.rst:
```suggestion
The GEOPM Runtime requires MPI standards, Message Passing Interface, version 2.2 or later. Fulfilling this requirement generally depends on the specific HPC resource targeted based on site-specific documentation. The Intel MPI implementation, OpenHPC or Spack packaging systems, or OpenMPI binaries distributed with most major Linux distributions satisfy this requirement. For RHEL and SLES Linux, the requirement can be met by installing the ``openmpi-devel`` package version 1.7 or later, and ``libopenmpi-dev`` on Ubuntu.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-intel_pstate kernel driver by modifying the kernel command line
-through grub2 or the boot loader on your system by adding:
+
+It's crucial to note that other Linux mechanisms for power management can interfere with GEOPM, which must be disenabled. Recommending disabling the intel_pstate kernel driver by modifying the kernel command line through grub2 or your system bootloader by adding:
```
```suggestion
It's crucial to note that other Linux mechanisms for power management can interfere with GEOPM, which must be disabled. Recommending disabling the ``intel_pstate`` kernel driver by modifying the kernel command line through grub2 or your system bootloader by adding:
```

Also, check grammar in second sentence.

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-to manually set the governor.  Older versions of SLURM require the
-desired governors to be explicitly listed in ``/etc/slurm.conf``.  In
-particular, SLURM 15.x requires the following option:
+`Cpufreq` driver will be enabled when the intel_pstate driver is disabled. The cpufreq driver has several modes controlled by the scaling_governor sysfs entry. When the performance mode is selected, the driver will not interfere with GEOPM. On SLURM-based systems, the :ref:`GEOPM launch wrapper <runtime:geopm application launch wrapper>` will attempt to set the scaling governor to "performance" a ...
```
```suggestion
The ``cpufreq`` driver will be enabled when the ``intel_pstate`` driver is disabled. It has several modes controlled by the ``scaling_governor`` sysfs entry. When the performance mode is selected, the driver will not interfere with GEOPM. On SLURM-based systems, the :ref:`GEOPM launch wrapper <runtime:geopm application launch wrapper>` will attempt to set the scaling governor to "performance" automatically, eliminating the need to manually set the governor. On older versions of SLURM, the desired governors must be listed explicitly in ``/etc/slurm.conf``. Specifically, SLURM 15.x requires the following option:
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-for information about how to share these implementations with the
-community).
+
+The GEOPM runtime requires each of the application's MPI processes to be affinitized to different CPUs. This is a critical requirement for the runtime and must be enforced by the MPI launch command. When using the geopmlaunch wrapper, these affinity requirements are handled by geopmlaunch when the ``--geopm-affinity-enable`` command-line option is provided (see :doc:`geopmlaunch(1)<geopmlaunch.1> ...
```
```suggestion
The GEOPM runtime requires each of the application's MPI processes to be affinitized to different CPUs. This is a critical requirement for the runtime and must be enforced by the MPI launch command. When using the ``geopmlaunch`` wrapper, these affinity requirements are handled by ``geopmlaunch`` when the ``--geopm-affinity-enable`` command-line option is provided (see :doc:`geopmlaunch(1)<geopmlaunch.1>`). Otherwise, users must explicitly affinitize their application using the appropriate options for their chosen launcher.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
+
+While the GEOPM control thread connects to the application it will automatically affinitize itself to the highest indexed core not used by the application if the application is not affinitized to a CPU on every core. If the application is using all cores of the system, the GEOPM control thread will be pinned to the highest logical CPU.
+
+Many ways exist to launch an MPI application, and no single uniform way of enforcing MPI rank CPU affinities can work across all job launch mechanisms. OpenMP runtimes, which are linked with compiler choice, also have different mechanisms for affinitizing OpenMP threads within CPUs available to each MPI process. The GEOPM control thread can also be launched as an application thread or process tha ...
```
```suggestion
Many ways exist to launch an MPI application, and no single uniform way of enforcing MPI rank CPU affinities can work across all job launch mechanisms. OpenMP runtimes, which are linked with compiler choice, also have different mechanisms for affinitizing OpenMP threads within CPUs available to each MPI process. The GEOPM control thread can also be launched as an application thread or process that can either be part of the primary MPI application or a completely different MPI application. Due to these factors, it is challenging to document the correct process affinitization across all configurations. Please refer to your site documentation about CPU affinity for the best solution for your system and consider extending the ``geopmlaunch`` wrapper to support your system configuration. For information on how to share these implementations with the community, refer to :doc:`contrib`.
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-resource manager via a SPANK plugin can be found in the `geopm-slurm git
-repository <https://github.com/geopm/geopm-slurm>`_. The implementation
-reflects what is documented below.
+The GEOPM Runtime package can seamlessly integrate with a compute cluster resource manager by altering the daemon of the resource manager running on the cluster compute nodes. An integration example with the SLURM resource manager through a SPANK plugin is available in the `geopm-slurm git repository <https://github.com/geopm/geopm-slurm>`_. This example aligns with the process described below.
```
Note: earlier in the documentation, the "Spank" capitalization was used, while here it's being capitalized as "SPANK". I would recommend using a consistent form throughout the documentation. Since it's an acronym, and seeing the Slurm documentation writes it as "SPANK", perhaps this form should be used.

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-path ``/etc/geopm/environment-override.json``:
+For GEOPM's integration as the universal power management solution for a cluster, it's usual for a single agent algorithm with one policy to be applied across all compute nodes within a partition. The choice of agent rests upon the site's needs. For instance, if the aim is to keep the average CPU power draw for each node below a specific cap, the :doc:`power_balancer agent <geopm_agent_power_bala ...
+
+To illustrate, if a system administrator wants to use the power_balancer agent, the process would involve setting a static power cap for apps not utilizing geopmlaunch, while optimizing power caps for performance when geopmlaunch is in use. The administrator would initiate the following JSON object in the compute node's boot image at ``/etc/geopm/environment-override.json``:
```
```suggestion
To illustrate, if a system administrator wants to use the ``power_balancer`` agent, the process would involve setting a static power cap for apps not utilizing ``geopmlaunch``, while optimizing power caps for performance when ``geopmlaunch`` is in use. The administrator would install the following JSON object in the compute node's boot image at ``/etc/geopm/environment-override.json``:
```

avilcheslopez commented on service/docs/source/runtime.rst:
```diff
@@ -261,271 +153,127 @@ option:
-the compute node boot image.  Changing the policy value will impact
-all subsequently launched GEOPM processes, but it will not change the
-behavior of already running GEOPM control processes.
+The controlling value, "CPU_POWER_LIMIT", is defined in a separate "geopm_power_balancer.json" file that could reside on a shared file system. This file can be generated using the :doc:`geopmagent(1) <geopmagent.1>` tool. By placing the policy file on a shared file system, you allow modifications to the limit without affecting the compute node boot image. Changing the policy value affects all new ...
```
```suggestion
The controlling value, ``CPU_POWER_LIMIT``, is defined in a separate "geopm_power_balancer.json" file that could reside on a shared file system. This file can be generated using the :doc:`geopmagent(1) <geopmagent.1>` tool. By placing the policy file on a shared file system, you allow modifications to the limit without affecting the compute node boot image. Changing the policy value affects all new GEOPM processes, but leaves running GEOPM processes untouched.
```

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -1,192 +1,93 @@
-architecture.  This information may be useful for system
-administrators, developers, and users who would like to better
-understand how the GEOPM Service provides security guarantees.
+This document explains the objectives, mechanisms, assets and threats associated with the design of the GEOPM Service security architecture. It can be beneficial to system administrators, developers, and users who wish to understand how the service provides security.
```
Column limit (more instances below)

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -1,192 +1,93 @@
-
-
-One writing GEOPM session allowed at any time
+The GEOPM Service provides a user gateway to privileged system software interfaces. Typically, these interfaces have Linux mechanisms that allow administrators to grant or deny user access. The service enhances other Linux security mechanisms by integrating fine-grained access management, easy extension, and an optional high-performance interface—improving upon what typically are coarse-grained,  ...
```
Check grammar around "...high-performance interface-improving..."

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -1,192 +1,93 @@
+
+Granting Linux capabilities to a helper command-line tool or adding a command-line tool to the `sudoers(5) <https://man7.org/linux/man-pages/man5/sudoers.5.html>`__ configuration is a common method of enabling specific secure features. These command-line tools may require independent security review and usually don't deliver high-performance programmatic solutions.
+
+High-performance security solutions often necessitate alterations to Linux kernel code either through a specialized driver or via `eBPF <https://ebpf.io>`__ program insertion. They require custom tools integrated into the Linux kernel, requiring independent security reviews by administrators and specialized kernel or eBPF development knowledge from developers and security experts. Identifying err ...
```
Check last sentence in this paragraph. Unless I'm having a hard time following, it seems something is off around "...in these filtering and device driver...".

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -1,192 +1,93 @@
+
+Users cannot make persistent changes to hardware settings through the GEOPM Service; any configuration alterations last only for the duration of the writing PID's process session. To understand the Linux process session interface more, refer to the `setsid(2) <https://man7.org/linux/man-pages/man2/setsid.2.html>`__ man page, which explains that a process session is usually linked to a controlling ...
+
+The GEOPM Service supports this feature by saving all control configuration settings to disk before enabling a user session to write. After the user's write session terminates, all configurations are restored to their original state. The saved state is stored in the directory ``/run/geopm/SAVE_FILES``.
```
```suggestion
The GEOPM Service supports this feature by saving all control configuration settings to disk before enabling a user session to write. After the user's write session terminates, all settings are restored to their original state. The saved state is stored in the directory ``/run/geopm/SAVE_FILES``.
```

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -206,326 +107,99 @@ JSON data.
-Intel Speed Select device driver
+The GEOPM Service can function as a gateway to the x86 Model-Specific Register device driver, `msr(4) <https://man7.org/linux/man-pages/man4/msr.4.html>`__, loaded as ``/dev/cpu/*/msr`` devices. There are various features accessible through the MSR device driver, and the GEOPM Service allows usage of a subset of these features focused on energy and power management and performance monitoring. Exa ...
+
+As direct access to the MSR driver may enable users to gain unauthorized information about processes they don’t own or influence system performance for other users, it's restricted. Using the MSR driver requires the CAP_SYS_RAWIO Linux `capability <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__. 
```
```suggestion
As direct access to the MSR driver may enable users to gain unauthorized information about processes they don’t own or influence system performance for other users, it's restricted. Using the MSR driver requires the ``CAP_SYS_RAWIO`` Linux `capability <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__. 
```

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -206,326 +107,99 @@ JSON data.
-
-
-LevelZero sysman library interface
+The specific features supported through the GEOPM Service are the `SST-CP <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-core-power-intel-r-sst-cp>`__ and `SST-TF <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-turbo-frequency-intel-r-sst-tf>`__ features. Use of the ``iss ...
```
```suggestion
The specific features supported through the GEOPM Service are the `SST-CP <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-core-power-intel-r-sst-cp>`__ and `SST-TF <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-turbo-frequency-intel-r-sst-tf>`__ features. Use of the ``isst_interface`` device driver necessitates the Linux `capability <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__ of ``CAP_SYS_ADMIN`` because changes may influence system performance for other users. The ISST interface can also alter the hardware characteristics reported by the Linux kernel, including the number of cores, base frequency, and achievable turbo frequencies.
```

avilcheslopez commented on service/docs/source/security.rst:
```diff
@@ -206,326 +107,99 @@ JSON data.
-interface.  The GEOPM Service relies on these standard Linux tools to
-provide a trusted interface and a secured attack surface.
-
+GEOPM Service leverages the systemd DBus, a standardized Linux interface for secure service communication. The `sd-bus(3) <https://man7.org/linux/man-pages/man3/sd-bus.3.html>`__ interface of the Linux systemd service enables secure request and result exchanges with the GEOPM Service, as well as the identification of client request origins. On the server side, the GEOPM DBus interface implementat ...
```
```suggestion
GEOPM Service leverages the systemd DBus, a standardized Linux interface for secure service communication. The `sd-bus(3) <https://man7.org/linux/man-pages/man3/sd-bus.3.html>`__ interface of the Linux systemd service enables secure request and result exchanges with the GEOPM Service, as well as the identification of client request origins. On the server side, the GEOPM DBus interface implementation utilizes the :doc:`dasbus <dasbus:index>` and :doc:`PyGObject <pygobject:index>` Python modules. In contrast, the client side employs ``libsystemd.so`` with the `sd-bus(3) <https://man7.org/linux/man-pages/man3/sd-bus.3.html>`__ interface. The GEOPM Service trusts these standard Linux tools for a reliable and secure interface.
```

avilcheslopez commented on service/docs/source/service.rst:
```diff
@@ -1,14 +1,9 @@
+User Guide for GEOPM Service
+=============================
 
+This user guide outlines the functionality and usage of the GEOPM Service. GEOPM Service is designed to provide a user interface for accessing hardware telemetry and settings. This guide encompasses sections detailing the process of building and installing this software component of the GEOPM project. Furthermore, it explains how to configure this service from an system administrator's standpoint ...
```
```suggestion
This user guide outlines the functionality and usage of the GEOPM Service. GEOPM Service is designed to provide a user interface for accessing hardware telemetry and settings. This guide encompasses sections detailing the process of building and installing this software component of the GEOPM project. Furthermore, it explains how to configure this service from a system administrator's standpoint and how an unprivileged user can interact with it.
```

avilcheslopez commented on service/docs/source/service_readme.rst:
```diff
@@ -7,71 +6,49 @@ Features
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-  Linux Systemd Service with DBus interface for user-level access to
-  hardware features on heterogeneous systems
+  The Linux Systemd Service is integrated with a DBus interface, enabling user-level access to hardware features on heterogeneous systems.
```
Column limit (more instances below)

avilcheslopez commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +60,39 @@ Architecture
-PlatformIO features.  The administrator uses the ``geopmaccess``
-command line tool to configure the DBus interface to enable user level
-access to any subset of the privileged PlatformIO features.
+The PlatformIO interface is a container for all IOGroups and is the main interface for users who interact with the hardware through the GEOPM Service. The PlatformIO interface can be accessed through language bindings with Python, C, and C++ as well as command line tools such as ``geopmread`` and ``geopmwrite``. The GEOPM DBus interface, ``io.github.geopm``, provides a secure gateway to privilege ...
```
```suggestion
The PlatformIO interface is a container for all IOGroups and is the main interface for users who interact with the hardware through the GEOPM Service. The PlatformIO interface can be accessed through language bindings with Python, C, and C++ as well as command line tools such as ``geopmread`` and ``geopmwrite``. The GEOPM DBus interface, ``io.github.geopm``, provides a secure gateway to privileged PlatformIO features. The ``geopmaccess`` command line tool is used by the administrator to enable user level access to any subset of the privileged PlatformIO features.
```

avilcheslopez commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +60,39 @@ Architecture
-useful for end-users, and it also provides information for system
-administrators that will help them understand what is enabled for an
-end-user when access is granted to a signal or control.
+Each signal and control in GEOPM has a unique name and a hardware domain. The signals and controls available on your system can be discovered with the ``geopmaccess`` command line tool. The description includes all necessary information for the end-users and system administrators for understanding what you enable an end user when you grant access to a signal or control.
```
```suggestion
Each signal and control in GEOPM has a unique name and a hardware domain. The signals and controls available on your system can be discovered with the ``geopmaccess`` command line tool. The description includes all necessary information for the end users and system administrators for understanding what you enable an end user when you grant access to a signal or control.
```

Check grammar in last sentence.

avilcheslopez commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +60,39 @@ Architecture
-adding a control name to the access list for writing, the
-administrator is implicitly providing read access to the control
-setting as well.
+All controls settings can be read by requesting the signal with the same name. Whenever a control name is added to the access list for writing, the administrator implicitly grants read access to the control setting as well.
```
```suggestion
All control settings can be read by requesting the signal with the same name. Whenever a control name is added to the access list for writing, the administrator implicitly grants read access to the control setting as well.
```

avilcheslopez commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +60,39 @@ Architecture
-that modify control values:
+Each time a client process opens a session with the GEOPM Service, a PlatformIO
+object is created with libgeopmd. This session starts in read-only mode. Calls
+to the D-Bus APIs that modify control values convert the session into write mode.
```
Inconsistent use of term "D-Bus". In most of the documentation, "DBus" is used.

avilcheslopez commented on service/docs/source/use_cases.rst:
```diff
@@ -1,54 +1,34 @@
-GEOPM comes with tools and software interfaces to access its
-features. This page outlines some of the use cases that are available
-across those interfaces.
+GEOPM bundles various tools and software interfaces designed for broad usability. This page outlines the multiple use cases catered for by these interfaces.
```
Column limit (more instances below)

avilcheslopez replied:
```suggestion
GEOPM bundles various tools and software interfaces designed for broad usability. This page outlines the multiple use cases catered by these interfaces.
```

avilcheslopez commented on service/docs/source/use_cases.rst:
```diff
@@ -1,54 +1,34 @@
-to choose optimal settings for hardware.  The algorithm used for optimization
-is a plugin interface that can be extended to reflect site or user
-requirements.
+A process can run on every compute node used by an High Performance Computing (HPC) workload. Leveraging hardware and software telemetry, this process optimally configures hardware settings. The optimization algorithm runs on a plugin interface, easily extensible to cater user or site-specific needs.
```
```suggestion
A process can run on every compute node used by a High Performance Computing (HPC) workload. Leveraging hardware and software telemetry, this process optimally configures hardware settings. The optimization algorithm runs on a plugin interface, easily extensible to cater to user or site-specific needs.
```

avilcheslopez commented on service/docs/source/use_cases.rst:
```diff
@@ -1,54 +1,34 @@
-Provide system administrators with a fine-grained access management interface
-that grants permissions to individual telemetry readings and control
-settings based on Linux user group membership.
+This feature equips system administrators with a granular access management interface. It allows them to allocate permissions for individual telemetry readings and control settings based either on Linux user group membership.
```
It seems this sentence is incomplete.

alawibaba committed:
Update documentation for 3.0.0 release.

Signed-off-by: Ali Mohammad <ali.mohammad@intel.com>

cmcantalupo commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,27 @@
-RPM packaging system enables some cross-Linux-distribution
-compatibility for the installation of the systemd service, and
-provides a clean mechanism for uninstall.
+We have tested packages created through this process on CentOS 8, RHEL 8, and
+SLES 15 SP2 Linux distributions. Feedback from installations on other
+distributions such as Fedora, openSUSE Leap, and Ubuntu would be greatly
```
update OS's tested.

cmcantalupo commented on service/docs/source/build.rst:
```diff
@@ -1,29 +1,27 @@
 
-Support for packaging for CentOS 8, RHEL 8, and SLES 15 SP2 is provided
-by the GEOPM service build system.
+The GEOPM service build system provides support for packaging for CentOS 8,
+RHEL 8, and SLES 15 SP2.
```
Update OS's

cmcantalupo commented on service/docs/source/build.rst:
```diff
@@ -45,37 +41,33 @@ directory:
     $HOME/rpmbuild/RPMS/x86_64/libgeopmd2-<VERSION>-1.x86_64.rpm
 
 
```
Add documentation for the `make deb` target.

cmcantalupo commented on service/docs/source/build.rst:
```diff
@@ -45,37 +41,33 @@ directory:
     sudo ./integration/install_service.sh $(cat VERSION) $USER
     systemctl status geopm
 
+Installation of the GEOPM Service Development Package
```
Should we update `install_service.sh` to support ubuntu?

cmcantalupo commented on service/docs/source/devel.rst:
```diff
@@ -1,197 +1,123 @@
+
+* ``--enable-nvml``: Adds support for the Nvidia NVML library
+* ``--enable-levelzero``: Adds support for OneAPI LevelZero
+* ``--disable-systemd``: Excludes GEOPM service access from PlatformIO
```
--disable-io-uring documentation missing, perhaps other new configuration options are also missing.

cmcantalupo commented on service/docs/source/docu.rst:
```diff
@@ -1,27 +1,25 @@
-Minimize the amount of whitespace (carriage returns, spaces, and indentation)
-present in the source file.  Only use the necessary amount of whitespace to
-satisfy the rst compiler for what you are trying to do.
+Strive to reduce the usage of whitespace (carriage returns, spaces, and indentation)
+in the source file. Use only the necessary amount of whitespace required by the rs compiler for your goal.
```
perhaps taken literally, this implies no \n.

cmcantalupo commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,86 @@
+
+Besides these routinely curated packages, the download repositories also grant
+access to ``python3-dasbus`` version 1.6. This version is a prerequisite
+for ``python3-geopmdpy``, but not included by the Linux distributions that
```
```suggestion
for ``python3-geopmdpy``, but not included by some of the Linux distributions that
```

cmcantalupo commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,86 @@
+Prerequisites for Level Zero
+----------------------------
+
+Before you install the GEOPM Service packages, ensure you have the
+``level-zero`` package. This package supports GEOPM for GPU metrics and
+controls. Currently, none of the Linux distributions supporting GEOPM
+provide this package upstream.
+
```
Level Zero dependent GEOPM packages are provided in the supplementary repo, but the default repo does not support Level Zero.  This should be clarified.

cmcantalupo commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,86 @@
+links will also guide you through instructions for each supported operating
+system. If you add both the development repository and the release repository
+to your package configuration, you will always receive updates from the
+development branch.
```
actually, I'm not sure this is true once a hot fix is created.  Once there is a hotfix, that version will be prefered by the package manager over the dev packages, e.g.: 2.0.1 > 2.0.0+dev999gabcdef1

cmcantalupo replied:
Advise the user add one repo or the other, not both.

cmcantalupo commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,86 @@
+What the Packages Don't Include
+-------------------------------
+
+Please note that the packages listed above do not offer the :doc:`GEOPM HPC
+Runtime <runtime>` features (e.g. :doc:`geopmlaunch(1) <geopmlaunch.1>`,
+:doc:`geopm_prof(3) <geopm_prof.3>`, :doc:`geopm_report(7) <geopm_report.7>`,
+``libgeopm.so`` and others.)
```
not true, MPI is disabled by default, but the runtime is enabled.

cmcantalupo commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,86 @@
+:doc:`geopm_prof(3) <geopm_prof.3>`, :doc:`geopm_report(7) <geopm_report.7>`,
+``libgeopm.so`` and others.)
+
+The GEOPM HPC Runtime should be built against a specific implementation of
+MPI, often unique to the system running the HPC application. An exception to
+this rule is the OpenHPC distribution. We've packaged GEOPM version 1 with
+`OpenHPC releases <http://openhpc.community/downloads/>`_, and we hope to
+offer version 2 in the future.
+
+The packages above don't provide support for NVIDIA GPUs with NVML or
+DCGM. This support depends on libraries not included in the OpenSUSE OBS
+build system where the GEOPM packages are curated. To enable the use of NVML,
+or DCGM, you will need to source-build the GEOPM Service, which requires
+an environment supporting the dependency libraries and headers.
```
Fact check this...

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -1,24 +1,19 @@
-
+To run the GEOPM service build, the packages listed below are required. All these
+packages are available in standard Linux distributions and can be installed
+using commands specific to RPM-based Linux distributions.
```
I think io-uring-devel is missing, and there should be a note about debian package support.

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -44,123 +39,99 @@ Upstream SLES and OpenSUSE Package Requirements
-currently packaged by Linux distributions (dasbus version 1.5 or more
-recent).  The script located in the subdirectory of the GEOPM repo:
+A more recent version of Dasbus (version 1.5 or later) is required for the GEOPM service.
+This version currently isn't packaged in Linux distributions. The GEOPM repository
+contains a script:
```
```suggestion
This version currently isn't packaged in some older Linux distributions. The GEOPM repository
contains a script:
```

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -44,123 +39,99 @@ Upstream SLES and OpenSUSE Package Requirements
 Dasbus Requirement
 ^^^^^^^^^^^^^^^^^^
```
Since dasbus 1.6 is available we should reduce or remove this documentation.

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -44,123 +39,99 @@ Upstream SLES and OpenSUSE Package Requirements
-sphinx.  This requirement may also be satisfied with PIP if installing the RPM
-packages for sphinx is an issue on your system:
+The Sphinx python package is essential for generating man pages and HTML
+documentation. For the build to function correctly, the man pages must be generated. These man pages are included in the distribution tarball created by the
```
nit...long line...

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -44,123 +39,99 @@ Upstream SLES and OpenSUSE Package Requirements
+also be managed by the system administrator using the GEOPM Service, if active.
+
+The msr-safe kernel driver code can be found `here <https://github.com/LLNL/msr-safe>`__.
+It's distributed with OpenHPC and can be installed from the RPMs provided there.
```
it is also distributed in the hardware repository of opensuse

cmcantalupo commented on service/docs/source/requires.rst:
```diff
@@ -44,123 +39,99 @@ Upstream SLES and OpenSUSE Package Requirements
-in ``/etc/systemd/logind.conf``.  Most Linux distributions change the
-default setting to disable this behavior.  More information can be
-found `here <https://superuser.com/a/1179962>`__.
+For GEOPM to utilize shared memory for communication between the Controller and the
+application, it may be necessary to alter systemd configuration. By default, systemd
+removes all inter-process communication for non-system users. This poses a problem to
+GEOPM's shared memory initialization routines.
```
This is not an issue any more.  The interprocess shared memory keys are managed in /run/geopm now.

cmcantalupo commented on service/docs/source/runtime.rst:
```diff
@@ -13,76 +12,65 @@ The architecture is designed to provide a secure infrastructure to
-
+For hardware tuning algorithms, the first challenge involves generating a
+durable estimate of application performance. In energy efficiency terms,
+performance measurements are power-to-performance ratios, for example,
```
```suggestion
performance measurements are performance-to-power ratios, for example,
```

cmcantalupo commented on service/docs/source/runtime.rst:
"requires" -> "provides optional support for"

in general should reword this section.

cmcantalupo commented on service/docs/source/security.rst:
```diff
@@ -1,192 +1,167 @@
+writing.  The client makes requests to read, write, or quit through
+a FIFO channel, connecting them to the batch server. A second FIFO
+communicates to the client when requests are complete. These FIFOs are
+opened in ``/tmp``, which systemd sets up by default as a `tmpfs(5)
+<https://man7.org/linux/man-pages/man5/tmpfs.5.html>`__.
```
these FIFOs are in /run/geopm.  I think Daniels correction to dev may have been clobbered.

cmcantalupo commented on service/docs/source/security.rst:
```diff
@@ -206,326 +181,216 @@ JSON data.
-information about the client session requests to be observed.
-
+To support GEOPM Service's batch server features, FIFO special files are
+created in the ``/tmp`` directory, working in tandem with inter-process
+shared memory. These FIFOs act as synchronization mechanisms, facilitating
```
these are not in /tmp but /run/geopm

cmcantalupo commented on service/docs/source/security.rst:
again, not /tmp

cmcantalupo commented on service/docs/source/service_readme.rst:
```diff
@@ -7,71 +6,68 @@ Features
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-  Users read telemetry from hardware components on heterogeneous
-  systems with a vendor agnostic interface
+  User can use a vendor-agnostic interface to read telemetry from hardware
```
```suggestion
  Use a vendor-agnostic interface to read telemetry from hardware
```

cmcantalupo commented on service/docs/source/service_readme.rst:
```diff
@@ -7,71 +6,68 @@ Features
-  Users may call a DBus interface to create a batch server which
-  provides a lower-latency interface with a single permissions
-  validation when server is created
+  A DBus interface can be utilized by the user for the creation of a batch
```
```suggestion
  A DBus interface can be utilized for the creation of a batch
```

cmcantalupo commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +79,57 @@ Architecture
+GEOPM Service. The administrator maintains an access list that applies to all
+users of the system. Special Unix groups can have enhanced access.
 
-   Before GEOPM 3.0, service configuration files were stored in
-   ``/etc/geopm-service``. Since version 3.0, they are stored in
-   ``/etc/geopm``. Version 3.0 ignores the old file location if the new
-   location exists. If the service uses a configuration from the old location,
-   then a deprecation warning is emitted.
-
-Any missing files are inferred to be empty lists, including the
-default access files.  A signal or control will not be available to
-non-root users through the GEOPM Service until a system administrator
-enables access through these allow lists.  It is recommended that all
-manipulation of these files should be done through the GEOPM Service
-with the ``geopmaccess`` command line tool.
-
-By convention, all control settings can be read by requesting the
-signal that shares the same name as the control.  Note that when
-adding a control name to the access list for writing, the
-administrator is implicitly providing read access to the control
-setting as well.
```
Not sure why this was deleted, should be brought back.

cmcantalupo commented on service/docs/source/service_readme.rst:
```diff
@@ -83,99 +79,57 @@ Architecture
-by the system administrator.  The administrator maintains a default
-access list that applies to all users of the system.  This list
-may be augmented so that users who are members of particular Unix groups may
-have enhanced access.  The default lists are stored in:
-
-.. code-block::
-
-   /etc/geopm/0.DEFAULT_ACCESS/allowed_signals
-   /etc/geopm/0.DEFAULT_ACCESS/allowed_controls
-
-
-Each Unix group name ``<GROUP>`` that has extended permissions can
-maintain one or both of the files
-
-.. code-block::
-
-   /etc/geopm/<GROUP>/allowed_signals
-   /etc/geopm/<GROUP>/allowed_controls
-
```
I think we still want these paths documented here...

cmcantalupo commented on service/docs/source/service_readme.rst:
```diff
@@ -200,16 +154,6 @@ restored to the value they had when the session was converted,
 regardless of whether or not they were adjusted during the session
 through the service.
 
-In addition to saving the state of controls, the GEOPM Service will
-also lock access to controls for any other client until the
-controlling session ends.  When the controlling session ends the saved
-state is used to restore the values for all controls supported by the
-GEOPM Service to the values they had prior to enabling the client to
-modify a control.  The controlling session may end by an explicit
-D-Bus call by the client, or when the process that initiated the
-client session ends.  The GEOPM Service will poll procfs for the
-process ID.
-
```
not sure why we want to remove this either.

cmcantalupo commented on service/docs/source/security.rst:
why were all the details removed from this section?

bgeltz committed:
Update build docu about supported OSes

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I6e64c0c160f4ac7bf540d7558b7f470c074a5c7d

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -1,197 +1,123 @@
+GEOPM Developer Guide
```
No discussion of the runtime RPMs and process.  Maybe that's OK here, but we should cook something up for the runtime.

bgeltz replied:
Open an issue to track refactoring the runtime.rst and devel.rst around OBS RPMs, Launchpad DEBs, and MPI support "requirements".

bgeltz committed:
Update developer guide

- Add new configure options.
- Add clang based compilers.
- Add textwidth style info.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: If8138df81f164fdd2911470d531bd4a114d5116b

bgeltz committed:
Update docu guidelines

- Add detail about line wrap columns.
- Align file to 80 columns.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: Id25d5d8b847f8b2b22d5c346dd10042b96c9c734

bgeltz committed:
Address issues found with docs_geopmlint

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I4e4fdbd91bef4c1ea2278a057dc51b0977544152

bgeltz committed:
Align section titles

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I967c53e040b45a14eadc65836b87f43b6c15b56a

bgeltz committed:
Update install guide

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: Ieb4347cc64eb3ce3a23813b5159285f083507fd5

bgeltz committed:
Updates to the runtime guide

- Address review feedback.
- Still too MPI centric.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I9e62ec3c799a72e3cb5c4f0f9b4f32bf231c1069

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -269,271 +237,236 @@ option:
-the user will explicitly affinitize their application using the appropriate
-options for their desired launcher.
+
+The GEOPM runtime requires each of the application's MPI processes to
```
This whole section still needs to be reworked.  Too MPI centric.

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -269,271 +237,236 @@ option:
+wrapper to support your system configuration. For information on how to
+share these implementations with the community, refer to :doc:`contrib`.
 
 Resource Manager Integration
```
Remove this section unless we're going to test it.

bgeltz commented on service/docs/source/install.rst:
```diff
@@ -1,115 +1,207 @@
+For information on how to install the GEOPM Runtime, see :doc:`GEOPM Runtime
+<runtime>`.
+
+.. MOVE THIS ALL TO A NEW RUNTIME INSTALLATION PAGE
+.. The GEOPM HPC Runtime should be built against a specific implementation of
+.. MPI, often unique to the system running the HPC application. An exception to
+.. this rule is the OpenHPC distribution. We've packaged GEOPM version 1 with
+.. `OpenHPC releases <http://openhpc.community/downloads/>`_, and we hope to
+.. offer version 2 in the future.
 
+.. The packages above don't provide support for NVIDIA GPUs with NVML or
+.. DCGM. This support depends on libraries not included in the OpenSUSE OBS
+.. build system where the GEOPM packages are curated. To enable the use of NVML,
+.. or DCGM, you will need to source-build the GEOPM Service, which requires
+.. an environment supporting the dependency libraries and headers.
```
Delete?

bgeltz commented on service/docs/source/build.rst:
```diff
@@ -40,36 +43,91 @@ These commands create the GEOPM service RPM files in your rpmbuild directory:
     $HOME/rpmbuild/RPMS/x86_64/python3-geopmdpy-<VERSION>-1.x86_64.rpm
     $HOME/rpmbuild/RPMS/x86_64/libgeopmd2-<VERSION>-1.x86_64.rpm
 
+Building GEOPM Service DEBs
```
Debian Packages

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -1,5 +1,5 @@
-GEOPM Developer Guide
-=====================
+Guide for GEOPM Developers
```
GEOPM Developer Guide

Also change the contributing guide to match (not here).

bgeltz commented on service/docs/source/docu.rst:
```diff
@@ -1,25 +1,38 @@
+
+Strive to minimize the usage of whitespace (carriage returns, spaces, and
+indentation) in the source file. Use only the necessary amount of whitespace
+as required for readability or by the rs compiler for your goal.
```
ReStructuredText

bgeltz commented on service/docs/source/runtime.rst:
```diff
@@ -340,7 +341,7 @@ application and the GEOPM control thread on each compute node, and manages
 :doc:`geopmlaunch(1)<geopmlaunch.1>` man page.
 
-``Geopmlaunch`` supports various underlying MPI application launchers
+``geopmlaunch`` supports various underlying MPI application launchers
```
The bla command

bgeltz commented on service/docs/source/install.rst:
```diff
@@ -11,76 +11,197 @@ process uses GitHub Actions, combined with the OpenSUSE Build Service
+.. code-block:: bash
+
+    # Add the upstream repo
+    $ add-apt-repository ppa:cmcantal/geopm-dev
```
ppa:geopm/dev
or
ppa:geopm/release

bgeltz committed:
Update install page ubuntu information

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I44f10147e468428610fe2ac364381ffe535778cc

bgeltz committed:
Address review feedback

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I0c77e3783f86b22ce8b2a518ffed7b40a0f1f07a

bgeltz committed:
Fix spelling errors

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I82991a32ebaf2a2a23b03cc5f0c32c311837adef

bgeltz committed:
Add OBS links for release-v3.0 configs

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: I0cf398519d856191944579130098579fe8cca3b1
